packages/cli/src/services/public-api-key.service.ts

```diff
@@ -134,17 +142,67 @@ export class PublicApiKeyService {
 		};
 	}
 
-	private generateApiKey = (user: User, expiresAt: UnixTimestamp) => {
+	private generateApiKey(user: User, expiresAt: UnixTimestamp) {
 		const nowInSeconds = Math.floor(Date.now() / 1000);
 
 		return this.jwtService.sign(
 			{ sub: user.id, iss: API_KEY_ISSUER, aud: API_KEY_AUDIENCE },
 			{ ...(expiresAt && { expiresIn: expiresAt - nowInSeconds }) },
 		);
-	};
+	}
 
 	private getApiKeyExpiration = (apiKey: string) => {
 		const decoded = this.jwtService.decode(apiKey);
 		return decoded?.exp ?? null;
 	};
+
+	apiKeyHasValidScopesForRole(role: GlobalRole, apiKeyScopes: ApiKeyScope[]) {
+		const scopesForRole = getApiKeyScopesForRole(role);
+		return apiKeyScopes.every((scope) => scopesForRole.includes(scope));
+	}
+
+	async apiKeyHasValidScopes(apiKey: string, endpointScope: ApiKeyScope) {
+		const apiKeyData = await this.apiKeyRepository.findOne({
+			where: { apiKey },
+			select: ['scopes'],
+		});
+		if (!apiKeyData) return false;
+
+		return apiKeyData.scopes.includes(endpointScope);
+	}
+
+	getApiKeyScopeMiddleware(endpointScope: ApiKeyScope) {
+		return async (req: Request, res: Response, next: NextFunction) => {
+			const apiKey = req.headers['x-n8n-api-key'] as string;
```

Comment:
```suggestion
			const apiKey = req.headers['x-n8n-api-key'];

			if ( apiKey === undefined || typeof apiKey !== 'string') {
				throw new UserError('something something')
			}
```

We can get rid of the casting here.

Reply:
The reason I did not do this it's because the `express-openapi-validator` will throw 401 in the header is not present. I guess to be extra safe in case we decide to use this middleware without the `express-openapi-validator` library.